### PR TITLE
vllm-omni framework and support for qwen2.5-omni [WIP]

### PR DESCRIPTION
**Pull request by Gaohan123**
- Reorganize worker modules: rename gpu_diffusion_* to new naming scheme
- Add new AR GPU worker and model runner
- Add diffusion scheduler and utilities
- Add stage configs for qwen2.5 omni model
- Update __init__.py files across modules for new imports
- Add output processor and arg utilities
- Add test file for omni LLM

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
based on my understanding, this huge PR can be split into the following mini PRs:

PR #1: Foundation & Infrastructure
Qwen3-Omni model configuration, loading, and tokenizer support
Scope: All foundational components needed for Qwen3-Omni support
Files:
vllm_omni/config/qwen3_omni_config.py (new)
vllm_omni/model_executor/model_loader/qwen3_omni_loader.py (new)
vllm_omni/model_executor/layers/qwen3_omni_tokenizer.py (new)
Update existing __init__.py files
Update vllm_omni/config/__init__.py
Changes:
Add Qwen3OmniConfig dataclass
Add model detection and loading logic
Add multimodal tokenizer wrapper
Add image preprocessing for Qwen3-Omni
Add configuration validation
Dependencies: None (can be merged independently)
Testing: Unit tests for all components

**Comment by hsliuustc0106** [comment]
PR #2: Core Processing Components
Model execution, request processing, and output handling
Scope: Core processing logic for Qwen3-Omni
Files:

1. vllm_omni/model_executor/models/qwen3_omni_model.py (new)
2. vllm_omni/request/qwen3_omni_request.py (new)
3. vllm_omni/engine/qwen3_omni_output_processor.py (new)
4. Update vllm_omni/request.py
5. Update vllm_omni/engine/output_processor.py

Changes:

1. Implement Qwen3OmniModel class with forward pass logic
2. Add Qwen3OmniRequest class with multimodal input validation
3. Implement Qwen3OmniOutputProcessor class
4. Add multimodal attention mechanisms
5. Add request preprocessing and output post-processing

Dependencies: PR #1
Testing: Unit tests for model execution, request processing, and output handling

**Comment by hsliuustc0106** [comment]
PR #3: Integration & API
Worker integration and API endpoints
Scope: Integration with existing vLLM-omni infrastructure
Files:

1. vllm_omni/worker/qwen3_omni_worker.py (new)
2. vllm_omni/entrypoints/qwen3_omni_api.py (new)
3. Update vllm_omni/worker/gpu_diffusion_worker.py
4. Update vllm_omni/entrypoints/api_server.py

Changes:

1. Add Qwen3-Omni worker class
2. Integrate with existing worker framework
3. Add GPU memory management
4. Add Qwen3-Omni specific API endpoints
5. Add request/response schemas
6. Add API documentation

Dependencies: PR #2
Testing: Integration tests and API tests

**Comment by hsliuustc0106** [comment]
PR #4: Examples & Documentation
Usage examples, configurations, and comprehensive documentation
Scope: Complete the Qwen3-Omni support with examples and docs
Files:

1. examples/qwen3_omni/ (new directory)
2. basic_usage.py
3. configs/qwen3_omni_local.yaml
4. README.md
5. docs/models/qwen3_omni.md (new)
6. Update docs/api/ with Qwen3-Omni endpoints
7. Update main README.md

Changes:

1. Add basic usage example
2. Add configuration examples
3. Add comprehensive model documentation
4. Add API documentation
5. Add troubleshooting guide
6. Update project documentation

Dependencies: PR #3
Testing: Example validation and documentation review

File: vllm_omni/entrypoints/omni_lm.py [hunk 1]
```diff
@@ -1,576 +1,140 @@
...
 
-        # Fix configuration validation issues
-        # Ensure max_num_batched_tokens is at least as large as max_model_len
-        if 'max_model_len' in kwargs and 'max_num_batched_tokens' in kwargs:
-            if kwargs['max_num_batched_tokens'] < kwargs['max_model_len']:
-                kwargs['max_num_batched_tokens'] = kwargs['max_model_len']
-        elif 'max_model_len' in kwargs:
-            # If max_model_len is set but max_num_batched_tokens is not, set it to max_model_len
-            kwargs['max_num_batched_tokens'] = kwargs['max_model_len']
+class OmniLM:
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/omni_lm.py:28]: what's the difference between OmniLM and OmniLLM

File: vllm_omni/entrypoints/omni_lm.py [hunk 2]
```diff
@@ -1,576 +1,140 @@
...
+        self.llm_engine.output_processor = MultimodalOutputProcessor(tokenizer=self.llm_engine.tokenizer, 
+                                                                    log_stats=self.llm_engine.log_stats)
+        self.llm_engine.processor = OmniProcessor(vllm_config=self.llm_engine.vllm_config,
+                                                  tokenizer=self.llm_engine.tokenizer)
+        self.engine_class = type(self.llm_engine)
+
+        self.request_counter = Counter()
+        self.default_sampling_params: Union[dict[str, Any], None] = None
+
+        if envs.VLLM_USE_V1:
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/omni_lm.py:147]: V0 has been removed

File: vllm_omni/engine/processor.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        params: Union[SamplingParams, PoolingParams],
+        arrival_time: Optional[float] = None,
+        lora_request: Optional[LoRARequest] = None,
+        tokenization_kwargs: Optional[dict[str, Any]] = None,
+        trace_headers: Optional[Mapping[str, str]] = None,
+        priority: int = 0,
+        data_parallel_rank: Optional[int] = None,
+    ) -> tuple[Optional[str], OmniEngineCoreRequest]:
+
+        # TODO(woosuk): Support pooling models.
```
Comment (hsliuustc0106) [inline vllm_omni/engine/processor.py:46]: do we copy this file from vllm?

File: vllm_omni/entrypoints/utils.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+from typing import List
+import os
+from omegaconf import OmegaConf
+from vllm.transformers_utils.config import get_config
+from vllm_omni.engine.arg_utils import OmniEngineArgs
+from vllm_omni.config import OmniStageConfig
+
+
+def load_stage_configs(omni_args: OmniEngineArgs) -> List[OmniStageConfig]:
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/utils.py]: load_stage_configs_from_engine_args

File: examples/offline_inference/qwen_2_5_omni/processing_omni.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+import torchvision
+from packaging import version
+from PIL import Image
+from torchvision import io, transforms
+from torchvision.transforms import InterpolationMode
+
+
+logger = logging.getLogger(__name__)
+
+IMAGE_FACTOR = 28
```
Comment (hsliuustc0106) [inline examples/offline_inference/qwen_2_5_omni/processing_omni.py:23]: do these magic numbers work for all models?

File: vllm_omni/worker/diffusion_model_runner.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (hsliuustc0106) [inline vllm_omni/worker/diffusion_model_runner.py:1]: for the model_runner and worker name, I think gpu_xx_model_runner and gpu_xx_worker are aligned with the AFD case: gpu_ffn_model_runner and gpu_ffn_worker.

File: vllm_omni/entrypoints/omni_lm.py (hunk 1 above)
Comment (Gaohan123) [reply vllm_omni/entrypoints/omni_lm.py:28]: OmniLM is a higher level abstraction as entry class, which init multiple OmniLLMs. Each OmniLLM inits one engine.

File: vllm_omni/entrypoints/omni_lm.py (hunk 2 above)
Comment (Gaohan123) [reply vllm_omni/entrypoints/omni_lm.py:147]: The code is based on a vLLM version around Aug 2025. Later we will adapt to latest stable vLLM v0.11.0

File: vllm_omni/engine/processor.py (hunk 3 above)
Comment (Gaohan123) [reply vllm_omni/engine/processor.py:46]: We modified some functions to add additional parameters.

File: vllm_omni/entrypoints/utils.py (hunk 4 above)
Comment (Gaohan123) [reply vllm_omni/entrypoints/utils.py]: Removed.

File: examples/offline_inference/qwen_2_5_omni/processing_omni.py (hunk 5 above)
Comment (Gaohan123) [reply examples/offline_inference/qwen_2_5_omni/processing_omni.py:23]: The file is for qwen2.5-omni only. It has been moved to examples.

**Comment by hsliuustc0106** [comment]
stale PR closed
